### PR TITLE
refactor compose

### PR DESCRIPTION
**Pull request by xenowits**
Refactors `docker-compose.yml` to remove port mappings (except for grafana) and also removes jaeger, loki and node-exporter. It moves the more "developer-focused" containers to a new "helper" compose file `compose-debug.yml`.

Also, updates the docker container versions.

ticket: https://github.com/ObolNetwork/charon-distributed-validator-node/issues/128

### CODE REVIEW

File: debug/.env.sample [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+#GETH_PORT_P2P=
+#GETH_PORT_HTTP=
+#GETH_PORT_AUTH=
+#GETH_PORT_METRICS=
+
+######### Lighthouse Config #########
+
+# Lighthouse beacon node docker container image version, e.g. `latest` or `v3.0.0`.
+# See available tags https://hub.docker.com/r/sigp/lighthouse/tags.
+#LIGHTHOUSE_VERSION=
```
Comment (corverroos) [inline debug/.env.sample:20]: lots of duplication here (with other .env file), not sure that is the best way to do it

File: debug/README.md [hunk 2]
```diff
@@ -0,0 +1,35 @@
+## Debug docker-compose.yml
+
+This folder contains a new `docker-compose.yml` file for developers to aid in debugging. This is meant to be configurable with sensible defaults that be overriden.
```
Comment (corverroos) [inline debug/README.md:3]: we should not duplicate the config, rather follow an additive approach where this folder adds to existing config.

**Review by corverroos — CHANGES_REQUESTED** [review]

File: compose-debug.yml [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+# Docker-compose file to aid developers in debugging. This is not required for "normal" users. See README.md for info.
+
+# Override any defaults specified by `${FOO:-bar}` in `.env` with `FOO=qux`.
+# See .env.sample "Debug Config" section
+
+# Execute "docker-compose -f docker-compose.yml -f compose-debug.yml up".
+
+services:
+  node-exporter:
```
[the remark this answers is not in the document]
Comment (xenowits) [reply compose-debug.yml:11]: yeah, but we can have it at the least until we add container metrics

**Comment by xenowits** [comment]
output of 
```sh
docker-compose -f docker-compose.yml -f compose-debug.yml up
```
Screenshot 2022-11-14 at 9 01 24 PM

File: README.md [hunk 4]
```diff
@@ -306,3 +313,10 @@ docker plugin install grafana/loki-docker-driver:latest --alias loki --grant-all
     - The charon metric `core_parsigdb_exit_total` will be incremented each time a voluntary exit partial signature is received, either from this node or from peers.
     - Once quorum partially signed voluntary exists have been received, they will be aggregated and submitted to the beacon node. This will add the validator to the beacon chain exit queue.
     - The validator keys can only be deleted from both `exit_keys` and `validator_keys` folders once the validator has successfully exited.
+
+13. I get an error `network charon-distributed-validator-node_dvnode declared as external, but could not be found` when I run `docker-compose -f docker-compose.yml -f compose-debug.yml up`.
+    - It occurs when docker can't find a network with the given name.
+    - Verify if the network is present by running:
+      - `docker network ls | grep charon-distributed-validator-node_dvnode`.
+    - If there are no results, simply create one:
+      - `docker network create charon-distributed-validator-node_dvnode`
```
Comment (corverroos) [inline README.md:322]: this is incorrect, the network prefix "charon-distributed-validator" comes from the folder name. Creating a network will not solve the issue. 

See above instructions:
```
      - Confirm the name of the exiting `charon-distributed-validator-node` docker network: `docker network ls`.
      - If it isn't `charon-distributed-validator-node-dvnode`, then update the `CHARON_DOCKER_NETWORK` env var in `.env` accordingly.
```

File: docker-compose.override.yml.sample [hunk 5]
```diff
@@ -17,16 +17,32 @@
 services:
   geth:
     # Disable geth
-    #profiles: [disable]
+    # profiles: [disable]
+    ports:
+      # - ${GETH_PORT_P2P:-30303}:30303/tcp # P2P TCP
```
Comment (corverroos) [inline docker-compose.override.yml.sample:22]: as I mentioned, no need to support overriding defaults in a file that is already an overide.

- Remove these overrides and provide hard coded values: `# - 30303:30303/tcp # P2P TCP`
- Remove the `GETH_PORT_*` override stubs in `.env.sample`

File: docker-compose.yml [hunk 6]
```diff
@@ -97,13 +80,8 @@ services:
...
       - CHARON_VALIDATOR_API_ADDRESS=0.0.0.0:3600
       - CHARON_MONITORING_ADDRESS=0.0.0.0:3620
-      - CHARON_JAEGER_ADDRESS=${CHARON_JAEGER_ADDRESS-jaeger:6831} # Overriding to empty address allowed
-      - CHARON_JAEGER_SERVICE=charon
-    ports:
-      - ${CHARON_PORT_VALIDATOR_API:-3600}:3600/tcp # Validator API
-      - ${CHARON_PORT_P2P_TCP:-3610}:3610/tcp       # P2P TCP libp2p
-      - ${CHARON_PORT_MONITORING:-3620}:3620/tcp    # Monitoring
-      - ${CHARON_PORT_P2P_UDP:-3630}:3630/udp       # P2P UDP discv5
+      - CHARON_LOKI_ADDRESSES=${CHARON_LOKI_ADDRESSES:-}
```
Comment (corverroos) [inline docker-compose.yml:83]: enable this by default, same as jaeger previously

Comment (corverroos) [reply docker-compose.yml:83]: `- CHARON_LOKI_ADDRESSES=${CHARON_LOKI_ADDRESSES-http://loki:3100/loki/api/v1/push} # Overriding to empty address allowed`

Comment (corverroos) [reply docker-compose.yml:83]: note the missing `:` in the `{$FOO-bar}` vs `{$FOO:-bar}`

File: .env.sample [hunk 7]
```diff
@@ -68,15 +68,29 @@
 # Monitoring host exposed ports
 #MONITORING_PORT_PROMETHEUS=
 #MONITORING_PORT_GRAFANA=
-#MONITORING_PORT_LOKI=
 #MONITORING_PORT_JAEGER=
 
+# Loki log aggregation server addresses
+#CHARON_LOKI_ADDRESSES=
```
Comment (corverroos) [inline .env.sample:74]: see comments below, but use the same comment as jaeger:
```
# Disable Loki log aggregation by setting an empty address.
#CHARON_LOKI_ADDRESSES=
```

File: compose-debug.yml [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+    image: bitnami/node-exporter:${NODE_EXPORTER_VERSION:-1.4.0}
+
+  jaeger:
```
Comment (corverroos) [inline compose-debug.yml:14]: enable jaeger in charon in this file:
```
services:
  charon:
    environment:
      - CHARON_JAEGER_ADDRESS=jaeger:6831
```

File: .env.sample (hunk 7 above)
Comment (corverroos) [reply .env.sample:74]: remove `CHARON_JAEGER_ADDRESS` above

File: docker-compose.yml [hunk 9]
```diff
@@ -52,16 +40,12 @@ services:
...
-      - ${LIGHTHOUSE_PORT_P2P:-9000}:9000/tcp   # P2P TCP
-      - ${LIGHTHOUSE_PORT_P2P:-9000}:9000/udp   # P2P UDP
-      - ${LIGHTHOUSE_PORT_METRICS:-5052}:5052   # Metrics
+    image: sigp/lighthouse:${LIGHTHOUSE_VERSION:-v3.2.1}
     command: |
       lighthouse bn
       --network=goerli
       --checkpoint-sync-url=https://goerli.checkpoint-sync.ethdevops.io
-      --execution-endpoint=http://geth:8551
+      --execution-endpoint=http://192.168.1.10:8551
```
Comment (corverroos) [inline docker-compose.yml:48]: hardcoded IP is not correct

File: docker-compose.yml [hunk 10]
```diff
@@ -162,35 +136,5 @@ services:
       - ./data/grafana:/var/lib/grafana
     restart: on-failure
 
-  loki:
```
Comment (corverroos) [inline docker-compose.yml:165]: do not remove loki, that was not part of the ticket

**Review by corverroos — CHANGES_REQUESTED** [review]

File: README.md (hunk 4 above)
Comment (xenowits) [reply README.md:322]: yeah, should set the network in .env, new steps:
```
    - It occurs when docker can't find a network with the given name.
    - Verify if the docker network is present by running:
      - `docker network ls | grep charon-distributed-validator-node_dvnode`.
    - If there are no results, update the `CHARON_DOCKER_NETWORK` env var in `.env` accordingly:
      - `CHARON_DOCKER_NETWORK=charon-distributed-validator-node_dvnode`
```

File: docker-compose.yml (hunk 6 above)
Comment (xenowits) [reply docker-compose.yml:83]: okay
```
${VARIABLE:-default} evaluates to default if VARIABLE is unset or empty in the environment.
${VARIABLE-default} evaluates to default only if VARIABLE is unset in the environment.
```

File: docker-compose.yml (hunk 10 above)
Comment (xenowits) [reply docker-compose.yml:165]: i was not sure as to include log aggregation as part of lean defaults. As discussed with the team, i'm adding back loki

File: README.md [hunk 11]
```diff
@@ -110,6 +111,9 @@ rm -r ./data/lighthouse
 # Spin up a Distributed Validator Node with a Validator Client
 docker-compose up
 
+# If you are a docker power user, run this instead:
+docker-compose -f docker-compose.yml -f compose-debug.yml up
```
Comment (corverroos) [inline README.md:115]: you need to include `docer-compose.overrides.yml` as well

File: README.md [hunk 12]
```diff
@@ -228,9 +232,10 @@ Keep checking in for updates, [here](https://github.com/ObolNetwork/charon/#supp
 
 5. How do I fix the `plugin "loki" not found` error?
```
Comment (corverroos) [inline README.md:233]: can remove this right?

File: README.md [hunk 13]
```diff
@@ -306,3 +312,10 @@ docker plugin install grafana/loki-docker-driver:latest --alias loki --grant-all
     - The charon metric `core_parsigdb_exit_total` will be incremented each time a voluntary exit partial signature is received, either from this node or from peers.
     - Once quorum partially signed voluntary exists have been received, they will be aggregated and submitted to the beacon node. This will add the validator to the beacon chain exit queue.
     - The validator keys can only be deleted from both `exit_keys` and `validator_keys` folders once the validator has successfully exited.
+
+13. I get an error `network charon-distributed-validator-node_dvnode declared as external, but could not be found` when I run `docker-compose -f docker-compose.yml -f compose-debug.yml up`.
+    - It occurs when docker can't find a network with the given name.
+    - Verify if the docker network is present by running:
+      - `docker network ls | grep charon-distributed-validator-node`.
+    - If there are no results, update the `CHARON_DOCKER_NETWORK` env var in `.env` accordingly:
+      - `CHARON_DOCKER_NETWORK=charon-distributed-validator-node_dvnode`
```
Comment (corverroos) [inline README.md:321]: this is confusing, your example is setting the default value, which is the problem at this point. Suggest: `CHARON_DOCKER_NETWORK=<network obtained from 'docker network ls'>`

File: docker-compose.override.yml.sample [hunk 14]
```diff
@@ -17,16 +17,45 @@
 services:
   geth:
```
Comment (corverroos) [inline docker-compose.override.yml.sample:18]: I get errors when using this:
 - ports cannot be empty
 - and subsequently the service-X cannot be empty 
 
 Suggest commenting out everything below `services`

File: README.md (hunk 12 above)
Comment (xenowits) [reply README.md:233]: yeah

File: docker-compose.override.yml.sample (hunk 14 above)
Comment (xenowits) [reply docker-compose.override.yml.sample:18]: yeah, the whole file needs to be commented out including `services:`

File: README.md (hunk 13 above)
Comment (xenowits) [reply README.md]: which network should be chosen if none exists? for ex: `docker network ls` gives:
```
NETWORK ID     NAME                            DRIVER    SCOPE
5cc259954324   493077459_compose               bridge    local
0291465e7045   1057849734_compose              bridge    local
7bbda699b8c3   2260650002_compose              bridge    local
00c47075ceb6   4023784422_compose              bridge    local
3e03359c3608   bridge                          bridge    local
0eb4ee0b17bd   charon-docker-compose_cluster   bridge    local
68154d9fc9c4   charon-release-06_compose       bridge    local
1f4205cc83c5   charon_validator                bridge    local
a9d469fcb9af   compose_compose                 bridge    local
61c81a1b6897   host                            host      local
a27a90dbb251   minikube                        bridge    local
5798ad791a45   none                            null      local
d0923f3970c5   soll_compose                    bridge    local
```

File: README.md [hunk 15]
```diff
@@ -110,6 +111,9 @@ rm -r ./data/lighthouse
 # Spin up a Distributed Validator Node with a Validator Client
 docker-compose up
 
+# If you are a docker power user, run this instead:
+docker-compose -f docker-compose.yml -f compose-debug.yml -f docker-compose.override.yml up
```
Comment (corverroos) [inline README.md:115]: swap override and debug

Comment (corverroos) [reply README.md:115]: and is this the only documentation you are adding for this compose-debug? It feels a bit light.

Comment (xenowits) [reply README.md:115]: gotcha, will add a new section for the "power users"

File: docker-compose.override.yml.sample [hunk 16]
```diff
@@ -15,18 +15,47 @@
...
+    # profiles: [disable]
+    # ports:
+      # - 9000:9000/tcp # P2P TCP
+      # - 9000:9000/udp # P2P UDP
+      # - 5052:5052 # Metrics
 
   charon:
     # Configure any additional env var flags in .env.charon.more
-    #env_file: [.env.charon.more]
+    env_file: [.env.charon.more]
```
Comment (corverroos) [inline docker-compose.override.yml.sample:38]: why uncomment this? this will break since users do not have this file.

File: README.md (hunk 15 above)
Comment (xenowits) [reply README.md:115]: would it be okay if I do it in a new PR?

Comment (corverroos) [reply README.md:115]: sure

Comment (xenowits) [reply README.md:115]: i'm gonna add the readme to this PR

File: README.md [hunk 17]
```diff
@@ -103,13 +104,18 @@ Before completing these instructions, you should assign a static local IP addres
...
+
 ```
 # Delete lighthouse data if it exists
 rm -r ./data/lighthouse
 
 # Spin up a Distributed Validator Node with a Validator Client
 docker-compose up
 
+# If you are a docker power user, run this instead:
+docker-compose -f docker-compose.yml -f compose-debug.yml -f docker-compose.override.yml up
```
Comment (corverroos) [inline README.md:117]: remove this (it belongs in the power user section)

File: compose-debug.yml [hunk 18]
```diff
@@ -0,0 +1,35 @@
+version: "3.8"
+
+# Docker-compose file to aid developers in debugging. This is not required for "normal" users. See README.md for info.
+
+# Override any defaults specified by `${FOO:-bar}` in `.env` with `FOO=qux`.
+# See .env.sample "Debug Config" section
+
+# Execute "docker-compose -f docker-compose.yml -f compose-debug.yml -f docker-compose.override.yml up".
```
Comment (corverroos) [inline compose-debug.yml:8]: order not correct

File: docker-compose.override.yml.sample (hunk 16 above)
Comment (corverroos) [reply docker-compose.override.yml.sample:38]: we can't include it either since it will cause fresh checkouts to fail. Why did it work previously?

**Review by corverroos — CHANGES_REQUESTED** [review]

File: README.md [hunk 19]
```diff
@@ -140,6 +146,34 @@ If you have gotten this far through the process, and whether you succeed or fail
...
+There are two additional files in this repository, `compose-debug.yml` and `docker-compose.override.yml.sample`, alongwith the default `docker-compose.yml` file for this purpose.
+
+- `compose-debug.yml` contains some additional containers that would developers can use for debugging, like `jaeger`. To achieve this, you can run:
+```
+docker-compose -f docker-compose.yml -f compose-debug.yml up
+```
+
+- `docker-compose.override.yml.sample` is intended to override the default configuration provided in `docker-compose.yml`. This is useful when, for example, you wish to add port mappings or want to disable a container.
+
+- We use the "Multiple Compose File" feature which provides a very powerful way to override any configuration in `docker-compose.yml` without needing to modify git-checked-in files since that results in conflicts when upgrading this repo.
```
[the remark this answers is not in the document]
Comment (xenowits) [reply README.md:159]: okay, makes sense!

File: compose-debug.yml (hunk 18 above)
Comment (xenowits) [reply compose-debug.yml]: removed this comment from this file, as the readme explains everything

File: docker-compose.override.yml.sample (hunk 16 above)
Comment (xenowits) [reply docker-compose.override.yml.sample]: i don't think it ever worked. i didn't use the override feature in this repo before so never encountered anything

**Review by corverroos — APPROVED** [review]

**Pull request merged**